### ISSUE TITLE
Column with subquery don't filtering

### ISSUE THREAD

**Issue by ASNOMO** [ThaDafinser/ZfcDatagrid#128]
Hello,

I have Column with sub query, but can not filter by this column, got PDO error, column name don't go to query.

$col = new Column\Select(new Expression("?", array($this->getServiceLocator()->get('vClassificatorsVatsTable')->getCountriesSubSelect())), 'country');
        $col->setLabel('Country');
        $col->setWidth(12);
        $grid->addColumn($col);

public function getCountriesSubSelect() {

```
    $select = new Select();
    $select->columns(array(new Expression("GROUP_CONCAT(c.title SEPARATOR ', ')")));
    $select->from(array('c' => self::TABLE_COUNTRIES));
    $select->join(array('vc' => self::TABLE_VATS_COUNTRY), 'vc.country_id = c.id', array(), 'left');
    $select->where('v.id = vc.vat_id');

    return $select;
}
```

**Comment by XCame** [comment]
@alihammad-gist @ThaDafinser 
I found a way to filter my subselect column which is using GROUP_CONCAT with Doctrine2 as dataSource.

I would label it as workaround but it's a working workaround ;)

Datasource:
```php
$qb = $entityManager->createQueryBuilder();
        $qb->select("company")
                ->from("ACME\Entity\ACMECompany", "company")
                ->leftJoin("company.organisation", "organisation") //same association as in subselect
                ->groupBy('company') //"distincts" the results
                ;

```
Column:
```php
 $col = new Column\Select(new Expr\Select('
            (
                SELECT
                    GROUP_CONCAT(
                        concatOrganisation.name SEPARATOR \'<br /> \'
                    )
                FROM ACME\Entity\ACMEOrganisation concatOrganisation
                WHERE
                    concatOrganisation.company = company
             )
        '), 'alias');
         
```

And now the magic:
```php
$col->setFilterSelectExpression('organisation.name');         
```

Why does it work?
Let's go the other way around; why didn't it work before?
When filtering the column with the subselect the following happens:

The Paginator wants to get the count so it builds a query like this 
```sql
SELECT COUNT(company) 
FROM company
JOIN organisation ON organisation.company = company
WHERE 
           (
                SELECT
                    GROUP_CONCAT(
                        concatOrganisation.name SEPARATOR '<br />'
                    )
                FROM ACME\Entity\ACMEOrganisation concatOrganisation
                WHERE
                    concatOrganisation.company = company
             )
 LIKE :theFilterInput
```

This one fails due to the WHERE.
You could replace the WHERE with HAVING then the LIKE will work and the Paginator gets its count.

But then you will face the next query related error, that the alias or class you are using in the subselect is already defined. 

That's because now the datagrid builds its query for getting the data
```sql
SELECT company.property1, 
               company.property2, 
               organisation.property1,
               organisation.property2,
              (
                SELECT
                    GROUP_CONCAT(
                        concatOrganisation.name SEPARATOR '<br />'
                    )
                FROM ACME\Entity\ACMEOrganisation concatOrganisation
                WHERE
                    concatOrganisation.company = company
             ) 
FROM object
JOIN joinedObject ON joinedObject.obj_id = object.id
WHERE         //or HAVING if you replaced it to make the Paginator work
              (
                SELECT
                    GROUP_CONCAT(
                        concatOrganisation.name SEPARATOR '<br />'
                    )
                FROM ACME\Entity\ACMEOrganisation concatOrganisation
                WHERE
                    concatOrganisation.company = company
             )
 LIKE :theFilterInput 
```

As you can see, the subselect with
```sql
....
FROM ACME\Entity\ACMEOrganisation concatOrganisation
....
```
is inserted twice and there you have you duplicate definition (of concatOrganisation).

Now
```php
$col->setFilterSelectExpression('organisation.name');         
```
is replacing the expression behind the 'WHERE' and before the 'LIKE :theFilterInput' with the string supplied.


Let's take a look what's happening with the Paginatior.
In now uses this to get the count
```sql
SELECT COUNT(company) 
FROM company
JOIN organisation ON organisation.company = company
WHERE organization.name LIKE :theFilterInput
```

seems pretty valid. 

And now the Datagrid's query for getting the Values:
```sql
SELECT company.property1, 
               company.property2, 
               organisation.property1,
               organisation.property2,
              (
                SELECT
                    GROUP_CONCAT(
                        concatOrganisation.name SEPARATOR '<br />'
                    )
                FROM ACME\Entity\ACMEOrganisation concatOrganisation
                WHERE
                    concatOrganisation.company = company
             ) 
FROM object
JOIN joinedObject ON joinedObject.obj_id = object.id
WHERE organisation.name LIKE :theFilterInput 
GROUP BY company //remember, we did this, when creating the queryBuilder
```

This also seems to be pretty valid ;-)




tl;dr;
- join the entity you use in the subselect with a different alias
- add groupBy() for your 'main' entity when creating the query builder
- replace the filterExpression of your subquery-column with the same property you're "GROUP_CONCAT"ing but with the joined entity as object.

Tadaaaaaaa :-)
